### PR TITLE
Self-hosted Engine CI for Windows and Linux

### PR DESCRIPTION
**Pull request by mwu-tow**
### Pull Request Description
[ci no changelog needed]

This PR adds a new workflow that builds engine on self-hosted Windows and Linux runners.
The build uses the new build script that is still under heavy development. Its snapshot that is known to work has been baked into the runner containers. As part of future tasks, the build script will be finished and integrated into this repository.

The purpose is twofold:
1.  Provide faster and more reliable CI for Engine on Windows.
2. Collect usage experience for the new build script and discover any potential issues.

As such, I'm leaving for now the traditional GitHub-hosted Engine CI intact, for comparison purposes and as a fallback. 

As soon as the PR is merged, I'll switch required PR checks from the old engine CI to the one introduced in this PR.

### Important Notes
This PR removes check for Rust version in `sbt`. It does not seem to work correctly in presence of the `rust-toolchain` file. And in general is spurious, as `rust-toolchain` should be enough to require a specific version.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/develop/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/develop/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

**Comment by 4e6** [comment]
@mwu-tow if we rely on `rust-toolchain` now, you should also update the rustup section here: https://github.com/enso-org/enso/blob/develop/docs/CONTRIBUTING.md?plain=1#L193-L203

```bash
rustup toolchain install nightly-2021-05-12
rustup override set nightly-2021-05-12
rustup component add clippy
```

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">enso/CONTRIBUTING.md at develop · enso-org/enso</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

**Comment by MichaelMauderer** [comment]
> @mwu-tow if we rely on `rust-toolchain` now, you should also update the rustup section here: https://github.com/enso-org/enso/blob/develop/docs/CONTRIBUTING.md?plain=1#L193-L203
> 
> ```shell
> rustup toolchain install nightly-2021-05-12
> rustup override set nightly-2021-05-12
> rustup component add clippy
> ```
> 
> >  GitHub**[enso/CONTRIBUTING.md at develop · enso-org/enso](https://github.com/enso-org/enso)**Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.

Yes, this needs updating. I thini we had this changed a whilw agon on the IDE side (pre-repo merge).

Also the version check could ensure that no local override is causing issues with the version selection. So there is some value in keeping it. 

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">enso/CONTRIBUTING.md at develop · enso-org/enso</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

**Comment by mwu-tow** [comment]
@4e6  @MichaelMauderer 
I've updates instructions as you suggested. Actually I noticed that there's a new form of rust-toolchain file that allows to define required components, so I used it and removed even more manual steps.

> Also the version check could ensure that no local override is causing issues with the version selection. So there is some value in keeping it.

Rust has an idiomatic way of saying "I need this specific version of toolchain" and this is rust-toolchain file. I would rather not keep it duplicated in our build scripts. Any override requires manual user action — in such case I assume it is intentional (e.g. to test a newer toolchain before bump).
As we discussed on the call, I agree that there's merit in build log including what version is actually used. I will ensure this happens as part of the new build script.

**Review by farmaazon — APPROVED** [review]
GUI parts looks ok.

File: rust-toolchain.toml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[toolchain]
```
Comment (akavel) [inline rust-toolchain.toml:1]: I'm somewhat confused what's the situation around clippy:
 - I don't see `components = ["clippy"]` here (is it not needed?)
 - I see `rustup component add clippy` removed from manual instructions
 - however, I ***still do*** see `rustup component add clippy` ***present*** in `.github/workflows/gui.yml`

I didn't test it, so maybe it works, just thought I'd ask if we're sure this is consistent?

Comment (mwu-tow) [reply rust-toolchain.toml:1]: @akavel You are right, thank you for catching this! Should be fixed now.

File: .github/workflows/scala-new.yml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+on:
+  push:
+    branches: ["*"]
+  pull_request:
+    branches: ["*"]
+  workflow_dispatch:
+
+jobs:
+  test_and_publish_win:
```
Comment (vitvakatu) [inline .github/workflows/scala-new.yml]: I think the name of the job is slightly inaccurate, as it has a `win` postfix while working on both Linux and Windows. It's not blocking and can be fixed in later PRs.

Other changes look nice to me.

**Review by vitvakatu — APPROVED** [review]

**Review by wdanilo — APPROVED** [review]

File: rust-toolchain.toml (hunk 1 above)
Comment (akavel) [reply rust-toolchain.toml:1]: sooo... still, what about the clippy installation entry in `.github/workflows/gui.yml`? 😅 can we remove it now?

**Review by 4e6 — APPROVED** [review]

**Pull request merged**